### DISCUSSION TITLE
Implementation of FEM with AMRex

### DISCUSSION THREAD

**Discussion by lw5366** [AMReX-Codes/amrex#3267] [General]
I am implementing an immersed boundary method with the Amrex framework, which has both a fluid flow solver and a solid solver. As the solid solver is based on finite element method, which raises a problem about how to handle the node and element with amrex, as they need to access each other during the computation.

The amrex provides classes to easily handle the "particles", which can be taken as the nodes in the solid solver. But, to solve the solid dynamics, we also need the element information (which stores something include the nodes numbers of this element, stress, strain and etc.). To achieve this, I am thinking to create two particlecontainers, with one of them to handle the nodes and another to handle the element. But it still has one issue, when we are handling the calculation in element, we need to access the nodes data as well, and vice versa. It is just like the particle needs to communicate with the background mesh which is inherhently supported by amrex. But here, the cooridnates of the two particlecontainers (i.e., node and element) are moving (Lagrangian description), which makes it not possible to utilise the inherent amrex feature to achieve the communication between them (in which the background mesh is Cartesian mesh and fixed).

To conclude, the issue is actually how to handle the communication between two particle containers, or is there any other easy way to achieve the FEM using amrex.

Thanks in advance for any suggestions.

**Comment by atmyers** [comment]
While you probably _could_ make something like this work with the AMReX particles, you'd need to implement a lot of functionality yourself (as you say, the ability to track which nodes belong to which element) and furthermore wouldn't be taking advantage of much of AMReX provides. If we ever do try to support FEM within AMReX, we'd do it by building new data structures that are designed for the task, rather than retrofitting it on top of the particles. So honestly I think you'd be better off exploring other packages - you might want to reach out to the [CEED](https://ceed.exascaleproject.org/) people and see if they have any advice.

**Reply by lw5366** [reply]
Thanks very much for your reply, and will have a look at the CEED.
Cheers!
